video/out/wayland_common.c

```diff
@@ -222,7 +222,10 @@ static void pointer_handle_button(void *data, struct wl_pointer *wl_pointer,
         button = MP_MBTN_FORWARD;
         break;
     default:
-        button = 0;
+        if (button >= BTN_FORWARD && button < BTN_JOYSTICK)
+            button += MP_MBTN9 - BTN_FORWARD;
+        else
+            button = 0;
```

Comment:
Honestly, I'm honestly confused at what this actually does. Does it really have to break the case statement?

Reply:
At the start of this switch statement, `button` is a `uint32_t` that—I assume—is expected to be somewhere in the range of Linux input event codes for mouse buttons, from 0x110 (`BTN_LEFT`) to 0x11F (doesn't have a name, but 0x120 is where the joystick event codes start). At the end of the switch statement, `button` needs to be a `uint32_t` that is either 0 or in the range of mpv key codes for mouse buttons, from 0x200000 (`MP_MBTN_LEFT`) to 0x200013 (`MB_MBTN19`). The mapping implemented so far is:

* 0x110 -> 0x200000 (`MP_MBTN_LEFT`)
* 0x111 -> 0x200002 (`MP_MBTN_RIGHT`)
* 0x112 -> 0x200001 (`MP_MBTN_MID`)
* 0x113 -> 0x200007 (`MP_MBTN_BACK`)
* 0x114 -> 0x200008 (`MP_MBTN_FORWARD`)
* everything else -> 0

The extra mappings I'm proposing adding are:

* 0x115 -> 0x200009 (`MP_MBTN9`)
* 0x116 -> 0x20000A (`MP_MBTN10`)
* 0x117 -> 0x20000B (`MP_MBTN11`)
* ...
* 0x11F -> 0x200013 (`MP_MBTN19`)

It's a simple linear mapping, so I thought the most straightforward thing to do would be to add the difference between the first button to be mapped on the mpv side (`MB_MBTN9`) and the first button to be mapped on the Linux side (0x115 is the unfortunately named `BTN_FORWARD`, which is not really the forward button) for any codes in the range to be mapped. If you'd rather see me add 11 case clauses to this switch instead, I can do that, I guess.

You could also object to the fact that codes above 0x117 are not currently defined in `input-event-codes.h`, which is true, but if at some point they're added, in my opinion this is the obviously correct thing to do with them, so I see no harm in mapping them now.

Reply:
See also this snippet from the Cocoa backend, where a similar catch-many linear mapping is applied at the end of a switch statement:

https://github.com/mpv-player/mpv/blob/8087e3371f8ffbdc8ddcbc0445392b0480ef02a4/video/out/cocoa/events_view.m#L334-L341

Reply:
I have no issues with adding these extra buttons, but I'd rather just see them explicitly added to the case statement (that's what it's for). The Cocoa implementation is nice and clever but it works since (presumably) cases 0 - 4 are all handled and anything other than that is 5 and up. Here, the default should always be 0 if we get some key we don't know; otherwise it gets confusing (well to me anyway). As for the codes above 0x117, since they are not actually defined, I do not think they should be added. If they ever get defined later, it another commit adding them to the mappings doesn't do any harm.

Reply:
Those extra codes aren't defined in the header, but that doesn't mean that the Linux HID driver doesn't emit them. For example, [this user](https://askubuntu.com/questions/19569/how-can-i-assign-actions-to-all-my-mouse-buttons) has a mouse with a button that Xorg/libinput maps to code 13, which corresponds to 0x118 in the kernel driver layer. ([Here's](https://gitlab.freedesktop.org/xorg/driver/xf86-input-libinput/-/blob/master/src/xf86libinput.c#L249-257) how the libinput driver for Xorg does that mapping. Hey, look, it's another linear map at the end of a switch statement!)

Look, you're the maintainer and I respect your right not to accept patches that you're not comfortable maintaining. But adding button mappings to the Wayland backend in small chunks *can* do harm. Specifically, it costs some future user who has more mouse buttons than I do the time it takes them to do all the same research I did when mpv doesn't recognize one of them. They need to figure out what in their input stack isn't working, from the hardware to the kernel module to the desktop environment to mpv itself. Then, upon correctly identifying that it's a failing in mpv, and perhaps after recognizing that in X, mpv picks up on the extra button [just fine](https://github.com/mpv-player/mpv/blob/073fbd98eeec86e62b8fada4fe9d42444981edfa/video/out/x11_common.c#L1203) (because the X layer has already done the mapping I linked above, it handles any event code that the HID driver produces, and mpv has no special cases to handle), they can find the same file I found, and submit a variation on the same patch I'm trying to submit now, which a maintainer can then spend more time reviewing. Why redo all that, when I've done the work and produced code that works now and in the future and brings feature parity across the Cocoa, X, and Wayland backends (I have no idea what W32 is doing)?

If your primary objection is just the nesting of an `if` inside a `switch`, fine, I'll write out the 11 case clauses. Or I'll write as much commentary as you think is necessary to make the linear shift comprehensible in your eyes—I don't see the difference between this and the Cocoa code (with Cocoa, cases below 5 are handled manually and otherwise handled with the shift; with this patch, cases below 0x115/`BTN_FORWARD` are handled manually and otherwise handled with the shift—is it just that `BTN_FORWARD` is a scarier constant than the number 5?). But I urge you to reconsider asking me to remove actual functionality and put the Wayland backend at a disadvantage relative to others.

Reply:
There seems to be some misunderstanding. I'm not against this. I know I said I didn't like buttons that weren't defined, but it wasn't a hard "no" (I can be persuaded). I just don't like the current implementation because it makes the switch statement look silly. When we hit the switch statement, we shouldn't be doing additional checks (yeah the nested `if` bothers me).

How about changing it so there's an `if/else` that happens outside of the switch? If the button is within the allowed range, then hit the switch statement (and you can cleanly add the linear map in there). If it's outside of it, then just make it 0. I know it's practically the same thing but the logic flow is clearer imo.

Reply:
Okay, that should work! Thanks.